File: LeetCode/198.House_Robber/Mario/198.House_Robber.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 0)
            return 0;
        if (n == 1)
            return nums[0];

        int[] res = new int[n];
        res[0] = nums[0];
        for (int i = 1; i < n; i++) {
            int j = i - 2;
            int k = i - 3;
            if (k >= 0) {
                res[i] = Math.max(res[j], res[k]) + nums[i];
            } else if (j == 0) {
                res[i] = res[j] + nums[i];
            } else if (j < 0) {
                res[i] = nums[i];
            }
        }
        return Math.max(res[n - 1], res[n - 2]);
    }
}

File: LeetCode/198.House_Robber/Mario/198.House_Robber.py
#!/usr/bin/env python3

class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)

        dp = [0] * (n + 1)

        # init dp
        dp[1] = nums[0]

        for i in range(2, n + 1):
            # rob current house
            cur = dp[i - 2] + nums[i - 1]
            # not rob current house
            pre = dp[i - 1]

            dp[i] = max(cur, pre)

        return dp[n]


"""
n = len(nums)
dp[n+1]
dp[i] robber can rob max amount of money from [1-i]
[1, 2, 3, 1]
[0, 1, 2, 4, 4]
"""